database/queries.go

```diff
@@ -0,0 +1,102 @@
+package database // import "go.mozilla.org/autograph/database"
+
+import (
+	"database/sql"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	// ErrNoSuitableEEFound is returned when no suitable key is found in database
+	ErrNoSuitableEEFound = errors.New("no suitable key found in database")
+)
+
+// BeginEndEntityOperations creates a database transaction that locks the endentities table,
+// this should be called before doing any lookup or generation operation with endentities.
+//
+// This global lock will effectively prevent any sort of concurrent operation, which is exactly
+// what we want in the case of key generation. Being slow and blocking is OK, risking two
+// key generation the happen in parallel is not.
+func (db *Handler) BeginEndEntityOperations() (*Transaction, error) {
+	// if a db is present, first create a db transaction to lock the row for update
+	tx, err := db.Begin()
+	if err != nil {
+		err = errors.Wrap(err, "failed to create transaction")
+		return nil, err
+	}
+	// lock the table
+	_, err = tx.Exec("LOCK TABLE endentities_lock IN SHARE MODE")
```

Comment:
I changed this to ACCESS EXCLUSIVE in a later revision to completely lock the table. This is effectively a global lock for all signers of this type, which I think is OK since we need to run this operations very rarely.